File: typescript/tests/construction_board.test.ts
// construction_board.test.ts
// Verifies the CircleAI.Construction port: open tasks by due date, complete,
// spend + remaining budget.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryConstructionBoard,
  ConstructionDomainContext,
  project,
  constructionTask,
  costEntry,
} from "../src/construction/index";

describe("InMemoryConstructionBoard", () => {
  it("lists open tasks for a project ordered by due date; complete closes them", () => {
    const b = new InMemoryConstructionBoard();
    b.create(project("p1", "House", new Date("2026-01-01T00:00:00Z"), null, 1_000_000, "ZAR"));
    b.add(constructionTask("t1", "p1", "Foundation", new Date("2026-02-01T00:00:00Z"), false));
    b.add(constructionTask("t2", "p1", "Roof", new Date("2026-01-15T00:00:00Z"), false));
    b.add(constructionTask("t3", "p1", "Paint", new Date("2026-03-01T00:00:00Z"), true)); // already done
    assert.equal(b.getProject("p1")?.name, "House");
    assert.deepEqual(
      b.openConstructionTasksFor("p1").map((t) => t.constructionTaskId),
      ["t2", "t1"],
    );
    b.complete("t2");
    assert.deepEqual(
      b.openConstructionTasksFor("p1").map((t) => t.constructionTaskId),
      ["t1"],
    );
  });

  it("complete throws on unknown task", () => {
    const b = new InMemoryConstructionBoard();
    assert.throws(() => b.complete("ghost"), /Unknown task ghost/);
  });

  it("sums spend and computes remaining budget; unknown project throws", () => {
    const b = new InMemoryConstructionBoard();
    b.create(project("p1", "House", new Date("2026-01-01T00:00:00Z"), null, 1000, "ZAR"));
    b.recordCost(costEntry("c1", "p1", "Materials", 300, new Date("2026-01-05T00:00:00Z")));
    b.recordCost(costEntry("c2", "p1", "Labour", 200, new Date("2026-01-06T00:00:00Z")));
    b.recordCost(costEntry("c3", "p2", "Other", 999, new Date("2026-01-06T00:00:00Z")));
    assert.equal(b.spendFor("p1"), 500);
    assert.equal(b.remainingBudget("p1"), 500);
    assert.throws(() => b.remainingBudget("ghost"), /Unknown project ghost/);
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(ConstructionDomainContext.systemPromptSnippet.includes("[DOMAIN: Construction]"));
    assert.deepEqual(ConstructionDomainContext.complianceFlags, [
      "OHS_Act",
      "NHBRC_Act",
      "CIDB_Act",
      "National_Building_Regs",
      "POPIA",
    ]);
    assert.deepEqual(ConstructionDomainContext.suggestedTools, ["project_scheduler", "document_editor", "map", "analytics"]);
  });
});

File: go/tests/construction_board_test.go
// construction_board_test.go
//
// Verifies the CircleAI.Construction port (construction_board.go): project
// create/get, task add/complete with open-task ordering, cost recording, spend
// totalling, and remaining-budget.

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestConstruction_TasksAndBudget(t *testing.T) {
	b := circleai.NewInMemoryConstructionBoard()
	start := time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC)
	b.Create(circleai.Project{ProjectId: "p1", Name: "House", StartOn: start, Budget: circleai.DecimalFromInt(100000), Currency: "ZAR"})
	if got, ok := b.GetProject("p1"); !ok || got.Name != "House" {
		t.Fatalf("get project = %+v ok=%v", got, ok)
	}

	b.Add(circleai.ConstructionTask{ConstructionTaskId: "t2", ProjectId: "p1", Description: "Roof", DueOn: start.AddDate(0, 0, 30)})
	b.Add(circleai.ConstructionTask{ConstructionTaskId: "t1", ProjectId: "p1", Description: "Foundation", DueOn: start.AddDate(0, 0, 10)})
	b.Add(circleai.ConstructionTask{ConstructionTaskId: "t3", ProjectId: "p1", Description: "Paint", DueOn: start.AddDate(0, 0, 60)})
	if err := b.Complete("t3"); err != nil {
		t.Fatalf("complete: %v", err)
	}
	if err := b.Complete("ghost"); err == nil {
		t.Fatalf("complete unknown must error")
	}
	open := b.OpenConstructionTasksFor("p1")
	if len(open) != 2 || open[0].ConstructionTaskId != "t1" || open[1].ConstructionTaskId != "t2" {
		t.Fatalf("open tasks ordered by DueOn failed: %+v", open)
	}

	now := time.Now().UTC()
	b.RecordCost(circleai.CostEntry{EntryId: "c1", ProjectId: "p1", Category: "materials", Amount: circleai.DecimalFromInt(30000), AtUtc: now})
	b.RecordCost(circleai.CostEntry{EntryId: "c2", ProjectId: "p1", Category: "labour", Amount: circleai.DecimalFromInt(20000), AtUtc: now})
	if spend := b.SpendFor("p1"); !spend.Equal(circleai.DecimalFromInt(50000)) {
		t.Fatalf("spend = %s, want 50000", spend.String())
	}
	rem, err := b.RemainingBudget("p1")
	if err != nil {
		t.Fatalf("remaining: %v", err)
	}
	if !rem.Equal(circleai.DecimalFromInt(50000)) {
		t.Fatalf("remaining budget = %s, want 50000", rem.String())
	}
	if _, err := b.RemainingBudget("ghost"); err == nil {
		t.Fatalf("remaining for unknown project must error")
	}
}
